### PR TITLE
fix(build): build type declarations separately

### PR DESCRIPTION
**Pull request by jinlee93**
[EFI-1103]

### Summary:
- builds declaration files separately since with the typescript rollup config the files exporting only types get pruned 

### Test Plan:

- [ ] Wrote [automated tests](https://czi.atlassian.net/wiki/x/Hbl1H)
- [ ] CI tests / new tests are not applicable
- [ ] Manually tested my changes, but I want to keep the details secret
- [x] Manually tested my changes, and here are the details:
  - Create an [alpha publish](https://github.com/chanzuckerberg/edu-design-system/blob/main/docs/PUBLISHING.md#alpha-release) and try out in `edu-stack` or `traject` as a sanity check if changes affect build or deploy, or are breaking, such as token changes, widely used component updates, hooks changes, and major dependency upgrades.
    - `v12.0.4-alpha.1`: Screenshot 2023-05-24 at 11 18 33 PM shows the `onChange` type being inherited properly and can navigate `node_modules/eds` to find the missing utility types in `edu-stack`
      - https://github.com/chanzuckerberg/edu-stack/tree/jlee/edsAlpha14.0.4

[EFI-1103]: https://czi-tech.atlassian.net/browse/EFI-1103?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Comment by jinlee93** [comment]
@ahuth Lmk if this helps with your ts issue at all

File: src/components/Input/Input.tsx [hunk 1]
```diff
@@ -37,13 +37,13 @@ export type InputProps = React.InputHTMLAttributes<HTMLInputElement> & {
    */
   isError?: boolean;
   /**
-   * Maximum number the input can take. When this number equals the input value, the plus button becomes disabled.
+   * Maximum number the input can take.
    */
-  max?: number;
+  max?: number | string | undefined;
```
Comment (booc0mtaco) [inline src/components/Input/Input.tsx:42]: `?:` might handle the `undefined` case already

**Review by booc0mtaco — COMMENTED** [review]
Can you say a bit more about what the package content looks like with this change applied?

File: src/components/Input/Input.tsx (hunk 1 above)
Comment (jinlee93) [reply src/components/Input/Input.tsx:42]: I just copy pasted `number | string | undefined` from `type HTMLInputElement` without seeing the `?:` haha will just simplify to `number | string`

File: package.json [hunk 2]
```diff
@@ -34,8 +34,9 @@
     "tailwind.config.*"
   ],
   "scripts": {
-    "build": "yarn build:clean && yarn build:tokens && yarn build:js && yarn copy-fonts-to-lib",
+    "build": "yarn build:clean && yarn build:tokens && yarn build:declarations && yarn build:js && yarn copy-fonts-to-lib",
     "build:clean": "rm -rf lib/",
+    "build:declarations": "tsc --emitDeclarationOnly --project tsconfig.build.json",
```
Comment (jeremiah-clothier) [inline package.json:39]: Does this also work with `--noEmit` be added to`build:js` portion?

**Comment by jinlee93** [comment]
> Can you say a bit more about what the package content looks like with this change applied?

Yup, looking at the file differences between building `next` and this branch, only adds `utility-types.d.ts`, which follows my intention of building definition files from `.ts` files that only export types. I tried converting `utility-types.ts` to the definition file but still gets pruned by rollup

File: package.json (hunk 2 above)
Comment (jinlee93) [reply package.json:39]: Rollup used in `build:js` also relies on `tsconfig.build.json` and having `noEmit` in there causes a logical error with building declarations `TS5053: Option 'emitDeclarationOnly' cannot be specified with option 'noEmit'.`

**Review by ahuth — APPROVED** [review]
LGTM

**Pull request merged**